clippy_lints/src/casts/cast_possible_truncation.rs

```diff
@@ -79,6 +80,21 @@ fn apply_reductions(cx: &LateContext<'_>, nbits: u64, expr: &Expr<'_>, signed: b
                 nbits
             }
         },
+        ExprKind::Path(ref _qpath) => get_constant_bits(cx, expr).unwrap_or(nbits),
+        // mem::size_of::<T>();
+        ExprKind::Call(func, []) => {
+            if let ExprKind::Path(ref func_qpath) = func.kind
+                && let Some(def_id) = cx.qpath_res(func_qpath, func.hir_id).opt_def_id()
+                && cx.tcx.is_diagnostic_item(sym::mem_size_of, def_id)
+                && let Some(ty) = cx.typeck_results().node_args(func.hir_id).types().next()
+                && let Ok(layout) = cx.tcx.layout_of(cx.param_env.and(ty))
+            {
+                let size = layout.layout.size();
+                size.bytes().next_power_of_two().trailing_zeros().into()
+            } else {
+                nbits
+            }
+        },
```

Comment:
`size_of` doesn't return the same value for every platform nor is a type's layout guaranteed to never change. If this were constrained to types defined in the current crate or any type with a known layout this would be fine.

Casting to a `u32` is probably also fine, but even a `u16` can be reasonably exceeded.

Reply:
> size_of doesn't return the same value for every platform
That's right. [But clippy is already affected by `cfg`. ](https://play.rust-lang.org/?version=nightly&mode=debug&edition=2021&gist=ab87ade127f1a2c107b522b5b6366a75)

> If this were constrained to types defined in the current crate or any type with a known layout

Could those be 
- primitive types
- std types, like String
- types in current crate
- not #[repr(Rust)]

Reply:
A struct from another crate has no guaranteed size since different versions of the crate may have different fields. Basically if a type is from another crate and is one of:
* A struct which is either non-exhaustive or contains a non-accessible field.
* A non-exhaustive enum.
* A union.

or it contains such a type, then we can't rely on it's size not changing. Some std types will have their own constraints that could be taken into account as well. (e.g.`Box<T>`, `ManuallyDrop<T>`)

Reply:
Okay, it all makes sense. I left a FIXME in [0f2daf3](https://github.com/rust-lang/rust-clippy/pull/12962/commits/0f2daf3207bea2e5308b7158c85a20a32095c1f3).